File: 2026/2nd_exer/Lights.java
import java.io.BufferedInputStream;
import java.io.FileInputStream;
import java.io.InputStream;

public class Lights {
	static InputStream in;
	static byte[] buffer = new byte[1 << 16];
	static int pos = 0, len = 0;

	static int read() throws Exception {
		if (pos == len) {
			len = in.read(buffer);
			pos = 0;
			if (len <= 0) return -1;
		}
		return buffer[pos++];
	}

	static int nextInt() throws Exception {
		int c = read();
		while (c <= ' ' && c != -1) c = read();

		int value = 0;
		while (c > ' ') {
			value = 10 * value + c - '0';
			c = read();
		}
		return value;
	}

	static void press(int[] rows, int n, int r, int c) {
		int bit = 1 << c;
		rows[r] ^= bit;
		if (c > 0) {
			rows[r] ^= 1 << (c - 1);
		}
		if (c + 1 < n) {
			rows[r] ^= 1 << (c + 1);
		}
		if (r > 0) {
			rows[r - 1] ^= bit;
		}
		if (r + 1 < n) {
			rows[r + 1] ^= bit;
		}
	}

	public static void main(String[] args) throws Exception {
		in = new BufferedInputStream(args.length > 0 ? new FileInputStream(args[0]) : System.in);

		int n = nextInt();
		int[] start = new int[n];
		for (int r = 0; r < n; r++) {
			for (int c = 0; c < n; c++) {
				if (nextInt() == 1) start[r] |= 1 << c;
			}
		}

		int best = Integer.MAX_VALUE;
		for (int firstRow = 0; firstRow < (1 << n); firstRow++) {
			int[] rows = start.clone();
			int presses = 0;

			for (int c = 0; c < n; c++) {
				if ((firstRow & (1 << c)) != 0) {
					press(rows, n, 0, c);
					presses++;
				}
			}

			for (int r = 1; r < n; r++) {
				int above = rows[r - 1];
				for (int c = 0; c < n; c++) {
					if ((above & (1 << c)) != 0) {
						press(rows, n, r, c);
						presses++;
					}
				}
			}

			if (rows[n - 1] == 0 && presses < best) best = presses;
		}

		if (best == Integer.MAX_VALUE) {
			System.out.println("IMPOSSIBLE");
		} else {
			System.out.println(best);
		}
	}
}

File: 2026/2nd_exer/lights.ml
let ic =
    if Array.length Sys.argv > 1 then open_in Sys.argv.(1) else stdin

let buffer = Bytes.create (1 lsl 16)
let len = ref 0
let pos = ref 0

let refill () =
    len := input ic buffer 0 (Bytes.length buffer);
    pos := 0

let read () =
    if !pos = !len then refill ();
    if !len = 0 then -1
    else
        let c = Char.code (Bytes.get buffer !pos) in
        incr pos;
        c

let next_int () =
    let c = ref (read ()) in
    while !c >= 0 && !c <= Char.code ' ' do
        c := read ()
    done;
    let value = ref 0 in
    while !c > Char.code ' ' do
        value := (10 * !value) + (!c - Char.code '0');
        c := read ()
    done;
    !value

let press a n i j =
    let bit = 1 lsl j in
    a.(i) <- a.(i) lxor bit;
    if j > 0 then a.(i) <- a.(i) lxor (1 lsl (j - 1));
    if j + 1 < n then a.(i) <- a.(i) lxor (1 lsl (j + 1));
    if i > 0 then a.(i - 1) <- a.(i - 1) lxor bit;
    if i + 1 < n then a.(i + 1) <- a.(i + 1) lxor bit

let () =
    let n = next_int () in
    let start = Array.make n 0 in

    for i = 0 to n - 1 do
        for j = 0 to n - 1 do
            if next_int () = 1 then start.(i) <- start.(i) lor (1 lsl j)
        done
    done;

    let best = ref max_int in

    (* try every possible first row *)
    for first = 0 to (1 lsl n) - 1 do
        let a = Array.copy start in
        let cnt = ref 0 in

        for j = 0 to n - 1 do
            if first land (1 lsl j) <> 0 then (
                press a n 0 j;
                incr cnt
            )
        done;

        for i = 1 to n - 1 do
            let above = a.(i - 1) in
            for j = 0 to n - 1 do
                if above land (1 lsl j) <> 0 then (
                    press a n i j;
                    incr cnt
                )
            done
        done;

        if a.(n - 1) = 0 && !cnt < !best then best := !cnt
    done;

    if !best = max_int then
        print_endline "IMPOSSIBLE"
    else
        Printf.printf "%d\n" !best
